Index: saunafs-4.5.1/configure
===================================================================
--- saunafs-4.5.1.orig/configure
+++ saunafs-4.5.1/configure
@@ -27,7 +27,6 @@ cmake .. \
 	-DENABLE_CLIENT_LIB=YES \
 	-DGSH_CAN_HOST_LOCAL_FS=YES \
 	-DENABLE_NFS_GANESHA=YES \
-	-DPACKAGE_VERSION="${version}" \
 	-DENABLE_URAFT=YES \
 	-DENABLE_POLONAISE=OFF \
 	-DCMAKE_CXX_FLAGS="-Wno-error=maybe-uninitialized"
